### PR TITLE
Check for Mac OS X 10.4 kqueue bug properly

### LINKED ISSUE #376: libevent erroneously detects kqueue bug on macOS 10.12
On macOS 10.12, libevent issues this warning and refuses to use kqueue:

[warn] kq_init: detected broken kqueue; not using.: Undefined error: 0

This detects an old bug in Mac OS X 10.4, fixed in 10.5. It would be surprising if that same bug resurfaced in 10.12.

The test for the bug is checking struct kevent::flags == EV_ERROR, but it should check struct kevent::flags & EV_ERROR.

We worked this at https://crbug.com/626534. This should be fixed in both the 1.x and 2.x series.

I don’t know if libevent intends to support anything as old as Mac OS X 10.4 at this point. Even if it does, the test and workaround can be moved into an #if that causes the code to not be built if it’s impossible for the code to run on an OS that old.

### PR DESCRIPTION
**Pull request by markmentovai**
`EV_ERROR` is a bit in `struct kevent::flags`. Other bits may be set too.

When it's not possible to run on Mac OS X 10.4, it's not necessary to test for the `kqueue` bug. If `MAC_OS_X_VERSION_MIN_REQUIRED` is the minimum runtime OS version that's being targeted. If it's set higher than 10.4, then compiled code is not expected to run on 10.4.

Fixes #376.

### CODE REVIEW

File: kqueue.c [hunk 1]
```diff
@@ -141,6 +145,7 @@ kq_init(struct event_base *base)
 		goto err;
 	kqueueop->events_size = kqueueop->changes_size = NEVENT;
 
+#if defined(__APPLE__) && MAC_OS_X_VERSION_MIN_REQUIRED < MAC_OS_X_VERSION_10_5
```
Comment (azat) [inline kqueue.c]: I don't think that we should compile it out, since it is `kq_init()` only, so let's leave 58bd1106e9728108ad8f91751e3bc00ea654d5ef

**Comment by azat** [comment]
I'm trying to find 10.12 to test this

**Comment by mark-chromium** [comment]
If you have a paid Apple Developer account, you can get the developer beta from https://developer.apple.com/download/. Otherwise, the public beta is at https://beta.apple.com/sp/betaprogram/.

**Comment by azat** [comment]
@mark-chromium  thanks, but firstly I need mac or vm.

**Comment by azat** [comment]
@mark-chromium I've just tested it, and I've picked the first one. Many thanks for a lot of details and a patch!

**Comment by mark-chromium** [comment]
Thanks, @azat. Are the 1.4 and 2.0 branches still live? Should this be applied there too? I can prepare pull requests for that if necessary.

**Comment by azat** [comment]
@mark-chromium if it is critical then I guess we can backport it to them, no need in pull requests I think, since patch should be picked without conflicts.

**Comment by mark-chromium** [comment]
I think it’s important enough to merge to any live branches. This bug effectively disables `kqueue()` support on macOS 10.12. `kqueue()` is normally the default backend on Macs as it’s the most functional and performant of the options, so not using it puts users of this recent system at a disadvantage.

In Chrome, we actually attributed a [hang bug](https://crbug.com/620259) to having `kqueue()` disabled. I haven’t dug deeper yet to determine what backend we were falling back to (probably `poll()`) or why we were missing events with that backend, but moving back to `kqueue()` as we had intended certainly made us stop missing events.

**Comment by azat** [comment]
Well I can't disagree with you, and this patch will backported to *2.0*, but I'm not sure about *1.4* (does anybody uses it?), except for chrome which have a copy of the whole library, but it is better to make it common instead of out-of-tree patches, so I guess that we can make some branch for this.

As for this bug, I don't have access to it, but it is interesting to look into it.

**Comment by mark-chromium** [comment]
Sorry about that! I’ve unrestricted the bug on our end.

**Comment by azat** [comment]
Thanks! BTW which location do you use to get **libevent 1.4** ?

**Comment by mark-chromium** [comment]
It looks like the last update pulled the tarball from [libevent.org](http://libevent.org/).

**Comment by TomerHeber** [comment]
Just ran into this bug when using "Sierra".
Thank you for handling it. I'll patch my Libevent and let you know if it fixed it.

**Comment by azat** [comment]
The backport landed as 6762a4d.

**Comment by azat** [comment]
And does somebody still uses 1.4? If so, is their any chance that they will migrate to 2.1 or at least 2.0?

**Pull request closed without merging**